### ISSUE TITLE
Cannot Upload Files to AWS S3

### ISSUE THREAD

**Issue by brian-weasner** [capacitor-community/http#179]
## Describe the bug

Unable to Upload Files to S3 within Android
Works within browser-served app.
Haven't tested in ios yet

## To Reproduce

1. Make a Post Request to save a file to an s3 bucket
2. Provide proper body as FormData with file as last attribute
  ```ts
  const url = `https://s3.amazonaws.com/${bucket}`;
  const formData = new FormData();
  formData.append('bucket', bucket);
  formData.append('key', uploadFileName);
  formData.append('acl', 'private');
  formData.append('Content-Type', file.type);
  formData.append('x-amz-server-side-encryption', 'AES256');
  
  formData.append('AWSAccessKeyId', accessKey);
  formData.append('Policy', policy);
  formData.append('Signature', signature);
  
  formData.append('file', file, uploadFileName);
  ```
3. Set 'Content-Type' header to 'multipart/upload'
    - Necessary for mobile to calculate boundary, breaks the browser because browser automatically knows it's multipart/upload based on FormData Object and then calculates Boundary
    - https://github.com/capacitor-community/http/blob/e675049b9b277bc4f4f0e6d446d107ebd183f447/android/src/main/java/com/getcapacitor/plugin/http/CapacitorHttpUrlConnection.java#L201

## Expected behavior

File to be uploaded

### [Success] Request from Browser/Served App
Browser/Served app's Network call:

**Request Headers:**
```
Accept: application/json, text/plain, */*
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cache-Control: no-cache
Connection: keep-alive
Content-Length: 2614966
Content-Type: multipart/form-data; boundary=----WebKitFormBoundaryBEHBJUvtT1jVkVDw
Host: s3.amazonaws.com
Origin: https://foo.bar:4200
Pragma: no-cache
Referer: https://foo.bar:4200/
sec-ch-ua: "Chromium";v="94", "Google Chrome";v="94", ";Not A Brand";v="99"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: cross-site
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36
```
**Form Data**
```
bucket: foo.bar.bucketname
key: 281_bdphbdokjx_1634163745230_dan-gold-N7RiDzfF2iw-unsplash.jpg
acl: private
Content-Type: image/jpeg
x-amz-server-side-encryption: AES256
AWSAccessKeyId: foobar
Policy: foobar
Signature: foobar
file: (binary)
```

### [FAIL] Request from Android

400 Error from Aws s3: POST requires exactly one file upload per request.

Android app's Network call:

**Request (from console log in inspector with remote debugging)**
```json
{
    "callbackId": "45448498",
    "pluginId": "Http",
    "methodName": "request",
    "options": {
        "method": "POST",
        "url": "https://s3.amazonaws.com/foo.bar.bucketname",
        "data": {},
        "headers": {
            "Content-Type": "multipart/form-data",
            "Accept": "application/json, text/plain, */*"
        },
        "params": {},
        "responseType": "text"
    }
}
```

**Request Headers from Android Network Profiler**
Note: the `Content-Type` header gets swapped out in the Android .java code to include the boundary
![image](https://user-images.githubusercontent.com/84519320/137223473-31270a16-f8b5-4b1c-b50d-f03360e43fa6.png)

**Response (from console log in inspector with remote debugging)**
```json
{
    "status": 400,
    "headers": {
        "Connection": "close",
        "Content-Type": "application/xml",
        "Date": "Wed, 13 Oct 2021 22:52:33 GMT",
        "Server": "AmazonS3",
        "Transfer-Encoding": "chunked",
        "x-amz-id-2": "foobar",
        "x-amz-request-id": "foobar",
        "X-Android-Received-Millis": "1634165554165",
        "X-Android-Response-Source": "NETWORK 400",
        "X-Android-Selected-Protocol": "http/1.1",
        "X-Android-Sent-Millis": "1634165554131"
    },
    "url": "https://s3.amazonaws.com/foo.bar.bucketname",
    "data": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>InvalidArgument</Code><Message>POST requires exactly one file upload per request.</Message><ArgumentName>file</ArgumentName><ArgumentValue>0</ArgumentValue><RequestId>foobar</RequestId><HostId>foobar</HostId></Error>",
    "error": true
}
```

**Response From Android Network Profiler**
![image](https://user-images.githubusercontent.com/84519320/137223896-08e52268-39be-4681-8047-99c09ae4addf.png)

## Device Details

**Desktop (please complete the following information):**
 - OS: Windows
 - Browser: Chrome
 - Version: 94

**Smartphone (please complete the following information):**
 - Device: Google Pixel 4
 - OS: Android 11
 - Browser: Chrome
 - Version: 83

## Additional Context/Debugging

- Research for 400 Error:
    - <https://stackoverflow.com/questions/37746434/amazon-s3-create-object-error-post-requires-exactly-one-file-upload-per-request>
    - <https://stackoverflow.com/questions/6030371/http-post-with-httpwebrequest>
        - Boundary Generation Seems Correct Here:
           - https://github.com/capacitor-community/http/blob/e675049b9b277bc4f4f0e6d446d107ebd183f447/android/src/main/java/com/getcapacitor/plugin/http/FormUploader.java
           - Boundary has 2 less `-` chars in the header than on every separation line, and last line of data has 2 `-` before and after the boundary
    - Differences between requests headers in browser and http plugin:
        - `Accept-Encoding: gzip, deflate, br`
        - `Accept-Language: en-US,en;q=0.9`
        - `Cache-Control: no-cache`
        - `Connection: keep-alive`
        - `Content-Length: 2614966`
            - Content-Length may be the issue?
            - Could set Content-Length with Java `HttpUrlConnection` via `setChunkedTransferMode`
                - <https://stackoverflow.com/questions/7062349/java-httpurlconnection-and-setting-the-content-length>
            - Tried setting file's Content-Length as another param in the FormData
                - Per S3 Rest Api Docs here: <https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax>
        - `Host: s3.amazonaws.com`
        - `Origin: https://foo.bar:4200`
        - `Pragma: no-cache`
        - `Referer: https://foo.bar:4200/`
    - Fiddler
        - Wanted to see if the request body had the boundary properly set in the body, but didn't see body in the chrome inspector or Android Profiler
        - Tried to intercept/read the requests that are being fired from the mobile app because Fiddler can decrypt https requests
        - Had issues and wasn't able to decrypt https requests from the app using this plugin, however was able to decrypt https requests made from the Chrome Browser on the device. (leading me to believe my configuration was correct)

**Comment by brian-weasner** [comment]
Is there a way to debug Native Java Code that is called by the plugin? I couldn't find any information online about how to debug native code that is called by a Capacitor Plugin?

**Comment by brian-weasner** [comment]
Used Pipedream service to see exactly what was being sent to S3:

NOTE: I'm using the plugin's `Http.request();` method instead of the plugin's `Http.uploadFile();` method, which I now know is causing me the trouble.

**Pipedream Event Log**
```json
{
    "method": "POST",
    "path": "/",
    "query": {},
    "headers": {
      "x-forwarded-for": "",
      "x-forwarded-proto": "https",
      "x-forwarded-port": "443",
      "host": "foo.m.pipedream.net",
      "x-amzn-trace-id": "Root=1-foobar",
      "content-length": "44",
      "accept-charset": "UTF-8",
      "accept": "application/json, text/plain, */*",
      "content-type": "multipart/form-data; boundary=1c6e6765-c95a-46ac-9cf7-8e25e7a693a3",
      "user-agent": "Dalvik/2.1.0 (Linux; U; Android 11; sdk_gphone_x86 Build/RSR1.201013.001)",
      "accept-encoding": "gzip"
    },
    "bodyRaw": "{}",
    "body": {}
  }
```
**NO BODY!!!!**

This is due to body being instance of FormData, so the Json object that is created from it looks empty.
Had to make a Json Object from From data

```ts
let data;
if (req.body instanceof FormData) {
  data = {};
  req.body.forEach(((value, key) => data[key] = value));
}
```

Made the request again.
**Pipedream Event Log**
```json
{
    "method": "POST",
    "path": "/",
    "query": {},
    "headers": {
      "x-forwarded-for": "",
      "x-forwarded-proto": "https",
      "x-forwarded-port": "443",
      "host": "foo.m.pipedream.net",
      "x-amzn-trace-id": "Root=1-foobar",
      "content-length": "2283",
      "accept-charset": "UTF-8",
      "accept": "application/json, text/plain, */*",
      "content-type": "multipart/form-data; boundary=b33b037f-2653-47ae-844f-b89dfcd78881",
      "user-agent": "Dalvik/2.1.0 (Linux; U; Android 11; sdk_gphone_x86 Build/RSR1.201013.001)",
      "accept-encoding": "gzip"
    },
    "bodyRaw": "{\"bucket\":\"foo.bar.bucketname\"}",
    "body": {
      "bucket": "foo.bar.bucketname"
    }
  }
```
**Not all the values of my form field...**
My js object made from my FormData has the following keys: `bucket, key, acl, Content-Type, Content-Length, x-amz-server-side-encryption, AWSAccessKeyId, Policy, Signature, file`

-- This is due to an error in `addFormField` method within `FormUploader.java`
    - adding end of request boundary to body after first/every field added to the request. (\`--${boundaryUid}--\`)
    - Switched the fn to use the same logic as `appendFieldToWriter` and I get the following:

**Pipedream Event Log**
```json
{
    "method": "POST",
    "path": "/",
    "query": {},
    "headers": {
      "x-forwarded-for": "",
      "x-forwarded-proto": "https",
      "x-forwarded-port": "443",
      "host": "foo.m.pipedream.net",
      "x-amzn-trace-id": "Root=1-foobar",
      "content-length": "1842",
      "accept-charset": "UTF-8",
      "accept": "application/json, text/plain, */*",
      "content-type": "multipart/form-data; boundary=a5845c3d-1a11-4958-9c37-42b8660a9143",
      "user-agent": "Dalvik/2.1.0 (Linux; U; Android 11; sdk_gphone_x86 Build/RSR1.201013.001)",
      "accept-encoding": "gzip"
    },
    "bodyRaw": "{\"bucket\":\"foo.bar.bucketname\",\"key\":\"133_la8ppwn9ur_1634246196078_tabea-schimpf-O7WzqmeYoqc-unsplash.jpg\",\"acl\":\"private\",\"Content-Type\":\"image/jpeg\",\"Content-Length\":\"3423018\",\"x-amz-server-side-encryption\":\"AES256\",\"AWSAccessKeyId\":\"foobar\",\"Policy\":\"foobar\",\"Signature\":\"foobar\",\"file\":\"{}\"}",
    "body": {
      "bucket": "foo.bar.bucketname",
      "key": "133_la8ppwn9ur_1634246196078_tabea-schimpf-O7WzqmeYoqc-unsplash.jpg",
      "acl": "private",
      "Content-Type": "image/jpeg",
      "Content-Length": "3423018",
      "x-amz-server-side-encryption": "AES256",
      "AWSAccessKeyId": "foobar",
      "Policy": "foobar",
      "Signature": "foobar",
      "file": "{}"
    }
  }
```
**My Js File Object is Empty!!!**

This is because when the `HttpRequestHandler` calls `JSValue data = new JSValue(call, "data");` the JSValue does not know how to convert the data object's file property into a "JSValue"/"JSObject".
It is for this reason that I believe that the plugin's `Http.uploadFile();` method exits, as it will take the path to the file and then read the file's contents to the request body within the native code, so no Js File Object conversion needs to happen.

**Comment by brian-weasner** [comment]
@thomasvidas  this is not fixed by Pr #178 
This issue has to do with how capacitor does not translate non primitive values (like an instance of Blob) from javascript to the native code. 
I do agree that this can be closed, because there is an upload file function, however the core reason as to why I created this issue/defect is not fixed by Url formatting.

**Comment by thomasvidas** [comment]
Whoops! My mistake @brian-weasner, a lot of the upload file issues were caused by URL formatting so I must've accidentally lumped this one in by mistake 😄
